sdk/core/System.ClientModel/src/ModelReaderWriter/Reflection/ReflectionModelBuilder.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.ClientModel.Primitives;
+
+[RequiresUnreferencedCode("This class uses reflection.  Pass in a generated ModelReaderWriterContext to be AOT compatible.")]
+internal class ReflectionModelBuilder : ModelReaderWriterTypeBuilder
+{
+    private Type _type;
+
+    protected override Type BuilderType => _type;
+
+    public ReflectionModelBuilder(Type type)
+    {
+        _type = type;
+    }
+
+    protected override object CreateInstance() => GetInstance(_type);
+
+    private static IPersistableModel<object> GetInstance(Type returnType)
+    {
+        var model = GetObjectInstance(returnType) as IPersistableModel<object>;
+        if (model is null)
+        {
+            throw new InvalidOperationException($"{returnType.ToFriendlyName()} does not implement {nameof(IPersistableModel<object>)}");
+        }
```

Comment:
nit: This can be simplified

```suggestion
        var model = GetObjectInstance(returnType) as IPersistableModel<object> ??
            throw new InvalidOperationException($"{returnType.ToFriendlyName()} does not implement {nameof(IPersistableModel<object>)}");
```